File: algorithms/3_hard/M/00076_minimum_window_substring/sliding_window.cpp
typedef ptrdiff_t isize;

class Solution {
public:
    static string minWindow(const string& s, const string& t) {
        const isize s_len = std::ssize(s);
        if (s_len < std::ssize(t)) { return string(); }
        isize freqs[256] = { 0 };

        for (const unsigned char ch : t) {
            freqs[ch] -= 1;
        }

        isize missing_distinct = 0;
        for (isize i = 0; i < 256; i += 1) {
            if (freqs[i] != 0) {
                missing_distinct += 1;
            }
        }
        isize lo = 0;
        auto res = span<const char>(&s[0], 0);

        for (isize hi = 0; hi < s_len; hi += 1) {
            freqs[static_cast<unsigned char>(s[hi])] += 1;
            if (freqs[static_cast<unsigned char>(s[hi])] == 0) {
                missing_distinct -= 1;
            }

            while (lo <= hi && freqs[static_cast<unsigned char>(s[lo])] > 0) {
                freqs[static_cast<unsigned char>(s[lo])] -= 1;
                lo += 1;
            }

            if (missing_distinct == 0 && (res.empty() || std::ssize(res) > (hi + 1 - lo))) {
                res = span<const char>(&s[lo], (hi + 1) - lo);
                // Remove leftmost required char to continue searching for a smaller window
                freqs[static_cast<unsigned char>(s[lo])] -= 1;
                lo += 1;
                missing_distinct += 1;
            }
        }

        return string(res.cbegin(), res.cend());
    }
};

File: algorithms/3_hard/M/00076_minimum_window_substring/sliding_window.rs
impl Solution {
    pub fn min_window(s: String, t: String) -> String {
        if s.is_empty() || t.is_empty() { return String::new(); }

        let s_bytes = s.as_bytes();
        let s_len = s_bytes.len();
        if s_len < t.len() { return String::default(); }
        let mut freqs: [isize; 256] = [0; 256];

        for &byte in t.as_bytes() {
            freqs[byte as usize] -= 1;
        }

        let mut missing_distinct = freqs.iter().filter(|&&x| x != 0).count();
        let mut lo = 0;
        let mut res = &s_bytes[0..0];

        for hi in 0..s_len {
            freqs[s_bytes[hi] as usize] += 1;
            if freqs[s_bytes[hi] as usize] == 0 {
                missing_distinct -= 1;
            }

            while lo <= hi && freqs[s_bytes[lo] as usize] > 0 {
                freqs[s_bytes[lo] as usize] -= 1;
                lo += 1;
            }

            if missing_distinct == 0 && (res.is_empty() || res.len() > (hi + 1 - lo)) {
                res = &s_bytes[lo..=hi];
                // Remove leftmost required char to continue searching for a smaller window
                freqs[s_bytes[lo] as usize] -= 1;
                lo += 1;
                missing_distinct += 1;
            }
        }

        String::from_utf8(res.to_vec()).unwrap()
    }
}
